OpenSim/Simulation/SimulationUtilities.h

```diff
@@ -0,0 +1,103 @@
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "osimSimulation.h"
+#include <simbody/internal/Visualizer_InputListener.h>
+
+namespace OpenSim {
+
+/**----------------------------------------------------------------------------
+Simulate a model from an initial state and return the final state.
+If the model's useVisualizer flag is true, the user is repeatedly prompted to
+either begin simulating or quit. The provided state is not updated but the
+final state is returned at the end of the simulation, when finalTime is reached.
+Set saveStatesFile=true to save the states to a storage file as:
+"<model_name>.sto". */
+inline SimTK::State simulate(Model& model,
+    const SimTK::State& initialState,
+    double finalTime,
+    bool saveStatesFile = false)
+{
+    // Returned state begins as a copy of the initial state
+    SimTK::State state = initialState;
+    SimTK::Visualizer::InputSilo* silo;
+
+    bool simulateOnce = true;
+
+    // Configure the visualizer.
+    if (model.getUseVisualizer()) {
+        SimTK::Visualizer& viz = model.updVisualizer().updSimbodyVisualizer();
+        // We use the input silo to get key presses.
+        silo = &model.updVisualizer().updInputSilo();
+
+        SimTK::DecorativeText help("Press any key to start a new simulation; "
+            "ESC to quit.");
+        help.setIsScreenText(true);
+        viz.addDecoration(SimTK::MobilizedBodyIndex(0), SimTK::Vec3(0), help);
+
+        viz.setBackgroundType(viz.GroundAndSky).setShowSimTime(true);
+        viz.drawFrameNow(state);
+        std::cout << "A visualizer window has opened." << std::endl;
+
+        // if visualizing enable replay
+        simulateOnce = false;
+    }
+
+    // Simulate until the user presses ESC (or enters 'q' if visualization has
+    // been disabled).
+    do {
+        if (model.getUseVisualizer()) {
+            // Get a key press.
+            silo->clear(); // Ignore any previous key presses.
+            unsigned key, modifiers;
+            silo->waitForKeyHit(key, modifiers);
+            if (key == SimTK::Visualizer::InputListener::KeyEsc) { break; }
+        }
+        else if (!simulateOnce) {
+            std::cout << "Press <Enter> to begin simulating, or 'q' followed "
+                << "by <Enter> to quit . . . " << std::endl;
+            if (std::cin.get() == 'q') { break; }
```

Comment:
If only simulating once, I think we do not need to wait for user input to start simulating, and we can avoid the logic around 'q'.